andy_elf/volmgr/volmgr.c

```diff
@@ -0,0 +1,779 @@
+            kputs("ERROR: volmgr_disk_write not yet implemented for PCI IDE\r\n");
+            return E_INVALID_DEVICE;
+        default:
+            kprintf("ERROR: volmr_disk_write invalid disk type 0x%x for 0x%x\r\n", disk->type, disk);
+            return E_INVALID_DEVICE;
+    }
+}
+
+void volmgr_vol_ref(struct VolMgr_Volume *vol) {
+    acquire_spinlock(&volmgr_lock);
+    vol->refcount++;
+    release_spinlock(&volmgr_lock);
+}
+void volmgr_vol_unref(struct VolMgr_Volume *vol) {
+    acquire_spinlock(&volmgr_lock);
+    if(vol->refcount>0) {
+        vol->refcount--;
+    } else {
+        kprintf("volmgr: Freeing volume 0x%x\r\n", vol);
+        free(vol->fs_ptr);
+        volmgr_disk_unref(vol->disk);
+        free(vol);
+    }
+    release_spinlock(&volmgr_lock);
+}
+
+void volmgr_disk_ref(struct VolMgr_Disk *disk) {
+    acquire_spinlock(&volmgr_lock);
+    disk->refcount++;
+    release_spinlock(&volmgr_lock);
+}
+
+void volmgr_disk_unref(struct VolMgr_Disk *disk) {
+    acquire_spinlock(&volmgr_lock);
+    if(disk->refcount>0) {
+        disk->refcount--;
+    } else {
+        for(struct VolMgr_Volume *vol = disk->volumes; vol!=NULL; vol=vol->next) {
+            kprintf("volmgr: Unref volume 0x%x from disk 0x%x\r\n", vol, disk);
+            volmgr_vol_unref(vol);
+        }
+        free(disk);
+    }
+    release_spinlock(&volmgr_lock);
+}
+
+void *volmgr_get_volume_by_name(const char *name) {
+    char maybe_marker = name[5];
+    if(maybe_marker != 'p' && maybe_marker != 'P') {
+        //Not a valid volume name
+        return NULL;
+    }
+    char maybe_partnum_char = name[6];
+    if(maybe_partnum_char < '0' || maybe_partnum_char > '9') {
+        //Not a valid volume name
+        return NULL;
+    }
+
+    uint8_t partnum = (uint8_t)(maybe_partnum_char - '0');
+    char disk_name[8];
+    strncpy(disk_name, name, 6);
+    struct VolMgr_Disk *disk = (struct VolMgr_Disk *)volmgr_get_disk_by_name(disk_name);
+
+    if(!disk) {
+        kprintf("volmgr: unable to find disk with name %s\r\n", disk_name);
+        return NULL;
+    }
+
+    acquire_spinlock(&volmgr_lock);
+    if(partnum >= disk->partition_count) {
+        kprintf("volmgr: requested partition %d but disk only has %d partitions\r\n", partnum, disk->partition_count);
+        release_spinlock(&volmgr_lock);
+        return NULL;
+    }
+    for(struct VolMgr_Volume *vol = disk->volumes; vol!=NULL; vol=vol->next) {
+        if(strncmp(vol->name, name, 8)==0) {
+            release_spinlock(&volmgr_lock);
+            return (void *)vol;
+        }
+    }
+    release_spinlock(&volmgr_lock);
+    return NULL;
+}
+
+uint8_t volmgr_disk_stash_pending_operation(struct VolMgr_Disk *disk, enum PendingOperationType type, void *buffer, void *extradata, uint16_t sector_count, uint64_t lba_address, void (*callback)(uint8_t status, void *buffer, void *extradata)) {
+    acquire_spinlock(&volmgr_lock);
+    for(size_t i=0; i<disk->pending_operation_count; i++) {
```

Comment:
Minor; we ignore `hand` and stash a pending op to the first available slot.  This can lead to unfair scheduling